### DISCUSSION TITLE
Use a custom instead issue's priority

### DISCUSSION THREAD

**Discussion by likehopper** [likehopper/redmine_sla#14] [Ideas]
Today, the definition of an « SLA term » is based on the « SLA type » and the Issue priorities enumeration ( present by in an issue natively ). Issue priority cannot be filtered by project and all statuses are always displayed.

**Comment by likehopper** [comment]
I identify three solutions at the moment:
- keep issue priority, the native enumeration field
- use a custom field which can be selected by SLA
- use a specific field of the plugin integrated in the issue

**Reply by maggne** [reply]
> Why not stay on three priorities (A, B & C) and sub-projects per customer (P1, P2, P3 and P4) distributed across four sub-projects (possibly each with its own SLA) ?

Answering to this:
It is much complicated, we've got x trackers, x projects, x customers. We need customers to not see priorities and SLAs amongst themselves. And they got SLAs set differently between each other.

So that's where it takes us. The custom field instead system field priority seems like the right way. It can be hidden for some users.
Will you work on it please?

**Reply by likehopper** [reply]
I am investigating the possibility of doing so.
But, for this to work, you would need a different custom field per SLA ?

**Reply by maggne** [reply]
Yes. Custom field created by us.

![obrazek](https://github.com/likehopper/redmine_sla/assets/157802929/214ae66a-b773-46e4-89ed-71845a51655d)

Here for example, you can see service level and customer priority, these are fields which defines the priority of the issue.
So, we are going to merge this two custom fields in to one, and the one would be field defining priority of issue and calculating SLA.
Is that possible to do?

**Reply by likehopper** [reply]
Yes, I'll try to offer the possibility to use a custom fields. But, it's going to take time. I think I will be able to come back to you in just a few weeks... Will that be okay?

**Reply by maggne** [reply]
Yes, take your time. I'm glad you are workng with us.
Are you going to release the API upgrade before? 
Please, let us know about every upgrade. We will be happy to test it out.
Thank you.

**Reply by maggne** [reply]
Hello, is there any update about that plugin? Would love to know all the upgrades you made. Thank you.

**Reply by likehopper** [reply]
The first implementation of APIs to query SLAs is progressing well.  I hope to commit a first version next week.  And then I will add other features based on this.

**Reply by zdenarik** [reply]
Hello @likehopper. I am contacting you instead of my colleague @maggne about progress of committing features you wrote above. Mainly APIs and possibility to set up SLA plugin by using custom fields. Thank you for your response

**Reply by likehopper** [reply]
1) I have just released a major new version:
- There is the implementation of management & API for sla_cache & sla_cache_spents with no longer use composite primary keys ( to use with Rails & Query ).
- In the interface, you can view the changes in the SLA admin panel. In the API documentation, at the bottom of the table, you will see the link to sla_cache and sla_cache_spent documentation. ( cf. https://github.com/likehopper/redmine_sla/blob/main/doc/API.md )

> **Required**: You must update the sla_caches and sla_cache_spents tables as well as all functions ( sla_get_date, sla_get_level_overlap, sla_get_level, sla_get_spent ) and all views ( sla_view_journal_statuses and sla_view_roll_statuses )

> **Note**: unit, system an functional tests has not yet been implemented for these parts.


2) I'm currently working on the next version for user-friendly API management (including knowing if deadlines are exceeded and possibly being able to send notifications via scripts). This is important to be able to truly use the SLA results to their full potential.


3) Only then I will try to offer the possibility to be able to choose between priorities or custom fields. I've started designing and it's not easy but I hope to get there as soon as possible.

Is this ok for you?

**Reply by zdenarik** [reply]
@likehopper Good work. It looks like good. Can I understand right that item sla_get_date is datetime field that shows deadline of SLA? Or that is field that shows datetime when SLA was created?
And one more question - how did you mean exactly things in note Required ? Shall I send this information to out Redmine admin and he will undestand? 
Thank you

**Reply by likehopper** [reply]
The item sla_get_date is datetime field shows when SLA actually started.

I'll see if it's possible to display a forecast end date, but it's difficult to determine. Because, if the status changes, and SLA goes back to "pause", the date becomes in an unknown state.

Yes, you can send the “required” note information to your administrator. In this note, I explain that it's necessary to rebuild two tables (sla_caches & sla_cache_spents), all the functions & all the views of the module.

**Reply by zdenarik** [reply]
OK, thank you. I see. 
And what about forecast end date  - when status changes to Waiting for Answer, forecast and date stay unchanged, Resolving time will stop and when it changes to In Progress for example, then it will add remaining time from Resolving time to forecast end date and updates it... is that possible?

**Reply by likehopper** [reply]
When the status changes to « Waiting for response », it is no longer possible to determine the « forecast end date ». It'll only be possible to determine the « forecast end date » when the status returns to « In Progress ».
In other words, if the status remains « Waiting for response », then there is nothing left to do ( only wait ) and there's therefore no more deadline.

**Reply by zdenarik** [reply]
OK, then. Thank you for explanation. Maybe in future.

**Comment by likehopper** [comment]
I've just committed the enhanced for use of a IssueCustomField in addition to IssuePriority. This week I will focus on testing and fixes. I therefore welcome your feedback!

These changes will need to be applied to the database after backup:

- ALTER TABLE sla_levels ADD COLUMN field custom_field_id BIGINT;
- ALTER TABLE sla_levels ADD CONSTRAINT sla_levels_custom_fields_fkey FOREIGN KEY (custom_field_id) REFERENCES public.custom_fields(id) ON DELETE CASCADE;
- ALTER TABLE sla_level_terms RENAME COLUMN priority_id TO priority;
- ALTER TABLE sla_level_terms ALTER COLUMN priority TEXT ;
- update sla_get_spent function

I'll modify the documentation only when everything is stabilized but you should quickly understand how to use the new functionality! Please feel free to ask me any questions to allow you to perform the tests.

**Reply by zdenarik** [reply]
Thank you for info. We will test these new features and let you know.

**Reply by likehopper** [reply]
I recommend downloading the latest version as it's enumeration based rather than list based.

**Comment by likehopper** [comment]
Have you been able to test the version I sent you? I just published a brand new stable version.

**Reply by zdenarik** [reply]
Hello @likehopper. Thank you for your interest. We are going to install and test version you sent us in next 10 days. I will inform as soon as we have finished testing.

**Reply by likehopper** [reply]
Please use the last version 🙏

**Reply by zdenarik** [reply]
Hello @likehopper. Finally, I have started to test you new version of Redmine SLA plugin after our Redmine admin installed things by you instructions. Unfortunatelly, some issues observed:

- there is missing button "Add SLA term" in SLA terms setting
![image](https://github.com/likehopper/redmine_sla/assets/168073013/b7309f67-863e-4cc9-a454-4847d92c93c6)
- after clicking on "SLA Caches" button I observe error message:
![image](https://github.com/likehopper/redmine_sla/assets/168073013/d836a392-f84d-43be-b14c-18e66630c5b7)
- after clicking on "SLA Caches spents" button I observe error message:
![image](https://github.com/likehopper/redmine_sla/assets/168073013/6f63f1a8-2efa-40b1-af1f-e472c98d47de)

Could you give me a hand with this?! O:-) Thank you

FYI
There is report with operations what our Redmine admin made:
- Update redmine_sla from actual main (root rm70427_sla_plugin_upgrade repa redmine_plugins)
- Build image (tag 5.0.8-ceos-test)
- Helm upgrade from appropriate image
- Next in DB:
> ALTER TABLE sla_levels ADD COLUMN custom_field_id BIGINT;

> ALTER TABLE sla_levels ADD CONSTRAINT sla_levels_custom_fields_fkey FOREIGN KEY (custom_field_id) REFERENCES public.custom_fields(id) ON DELETE CASCADE;

> ALTER TABLE sla_level_terms RENAME COLUMN priority_id TO priority;

> ALTER TABLE sla_level_terms ALTER COLUMN priority set data type text;

And he has no idea what is "update sla_get_spent function". He suppose that plugin should apply this itself. Is that true?

**Reply by likehopper** [reply]
I apologize for these inconveniences. I will prioritize looking into this to fix it and add the associated tests to avoid regressions.
I would appreciate receiving the export of the database structure (with the --schema-only option).

**Reply by zdenarik** [reply]
OK, nevermind... I have contacted our Redmine admin and asked for that...

**Reply by zdenarik** [reply]
Hello @likehopper. I am very sorry, but I cannot provide the export of database because of busyness of our Redmine Admin and time consumption difficulty for do that. Are you OK with this? Thank you for understanding

**Reply by likehopper** [reply]
Yes, if you can just apply this script to rebuild the tables and functions : [redmine_sla_move_from_priority_to_customfield.zip](https://github.com/user-attachments/files/15597278/redmine_sla_move_from_priority_to_customfield.zip).

**Comment by likehopper** [comment]
**there is missing button "Add SLA term" in SLA terms setting**
In fact, it's no longer possible to add a term since the priority name depends on the custom field. You need to go to the SLA Level Index and click on the first action “Edit SLA Term” : 
![image](https://github.com/likehopper/redmine_sla/assets/12208142/6aa91e28-1d59-4955-9d14-6aa8e0da8736)

**Reply by zdenarik** [reply]
OK, I will try that... thx

**Comment by likehopper** [comment]
While waiting for the database structure to confirm, I think the problem is with the tables. It would be best to apply the attached script to rebuild the tables and functions : [redmine_sla_move_from_priority_to_customfield.zip](https://github.com/user-attachments/files/15597278/redmine_sla_move_from_priority_to_customfield.zip).
Finally, it is possible to run the script to force the cache update ( `rake redmine:plugins:redmine_sla:update_sla` ). Additionally, the best is to download the latest version of the code (while waiting for its stable publication and versioning).

**Reply by zdenarik** [reply]
Handed over to Redmine admin... thx

**Comment by likehopper** [comment]
Have you tested the latest available version ? Does it meet your needs ?

**Reply by zdenarik** [reply]
Hello @likehopper. Thank you for you interest. Sorry, but we have not tested it. We have more important project duties to be done. Nevertheless I am going to test it during summer holidays when our people will be on vacation. I will let you know. Thank you again for your understanding.
